File: DSA/pranavdixit_590014418/Day_25_Question_2.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* newNode(int val) {
    struct Node* temp = (struct Node*)malloc(sizeof(struct Node));
    temp->data = val;
    temp->next = NULL;
    return temp;
}

void append(struct Node** head, int val) {
    struct Node* temp = newNode(val);
    if (*head == NULL) {
        *head = temp;
        return;
    }
    struct Node* curr = *head;
    while (curr->next)
        curr = curr->next;
    curr->next = temp;
}

struct Node* getIntersection(struct Node* a, struct Node* b) {
    struct Node* result = NULL;
    struct Node* tail = NULL;

    while (a && b) {
        if (a->data == b->data) {
            if (result == NULL) {
                result = newNode(a->data);
                tail = result;
            } else {
                tail->next = newNode(a->data);
                tail = tail->next;
            }
            a = a->next;
            b = b->next;
        } else if (a->data < b->data) {
            a = a->next;
        } else {
            b = b->next;
        }
    }

    return result;
}

void printList(struct Node* head) {
    while (head) {
        printf("%d ", head->data);
        head = head->next;
    }
    printf("\n");
}

int main() {
    struct Node* list1 = NULL;
    struct Node* list2 = NULL;

    append(&list1, 1);
    append(&list1, 2);
    append(&list1, 3);
    append(&list1, 4);
    append(&list1, 6);

    append(&list2, 2);
    append(&list2, 4);
    append(&list2, 6);
    append(&list2, 8);

    printf("List 1: ");
    printList(list1);

    printf("List 2: ");
    printList(list2);

    struct Node* intersect = getIntersection(list1, list2);

    printf("Intersection: ");
    printList(intersect);

    return 0;
}

File: DSA/PranyaGupta_500126032/Day25Question2.py
class node:
    def __init__(self, data):
        self.next = None
        self.data = data
class LinkedList:
    def __init__(self):
        self.head = None
    def add(self, data):
        new_node = node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node
    def printlist(self, head):
        current = head
        while current:
            print(current.data, end=" ")
            current = current.next        
    
    def intersection(self, head1, head2):
        curr1 = head1
        curr2 = head2
        res = LinkedList()
        while(curr1 and curr2):
            if curr1.data == curr2.data:
                res.add(curr1.data)
                curr2= curr2.next 
                curr1 = curr1.next 
            elif curr1.data<curr2.data:
                curr1 = curr1.next
            else:
                curr2 = curr2.next        
        self.printlist(res.head)    
            

def main():
    user_input1 = input("Enter elements of the linked list1 (with space in between): ")
    arr1 = list(map(int, user_input1.strip().split()))
    l1 = LinkedList()
    for n1 in arr1:
        l1.add(n1)
    user_input2 = input("Enter elements of the linked list2 (with space in between): ")
    arr2 = list(map(int, user_input2.strip().split()))
    l2 = LinkedList()
    for n2 in arr2:
        l2.add(n2)
    l = LinkedList()   
    l.intersection(l1.head1, l2.head2)
if __name__ == "__main__":
    main()